synapse/storage/databases/main/events_worker.py

```diff
@@ -1356,14 +1356,23 @@ async def have_seen_events(
         Returns:
             The set of events we have already seen.
         """
-        res = await self._have_seen_events_dict(
-            (room_id, event_id) for event_id in event_ids
-        )
-        return {eid for ((_rid, eid), have_event) in res.items() if have_event}
+
+        # @cachedList chomps lots of memory if you call it with a big list, so
+        # we break it down. However, each batch requires its own index scan, so we make
+        # the batches as big as possible.
```

Comment:
OOI do we know *why* `@cachedList` is so heavyweight? That feels like it might be a useful thing to fix (separately)

Reply:
I *think* it's because it creates a new Deferred for each (missing) entry in the list. Maybe?

I agree it would be a useful thing to investigate. I did consider trying to do add support for batching to `@cachedList`, but it looked scary.